src/reference.js

```diff
@@ -1224,19 +1231,64 @@ class Query {
    * Builds a Firestore 'Position' proto message.
    *
    * @private
-   *
-   * @param {Array.<*>} fieldValues - The set of field values to use as the
-   * boundary.
+   * @param {Array.<DocumentSnapshot|*>} fieldValues - The snapshot of the
+   * document or the set of field values to use as the boundary.
    * @param before - Whether the query boundary lies just before or after the
    * provided data.
    * @returns {Object} The proto message.
    */
-  _buildPosition(fieldValues, before) {
+  _buildPosition(fieldValuesOrDocumentSnapshot, before) {
     let options = {
       before: before,
       values: [],
     };
 
+    let fieldValues;
+    if (
+      fieldValuesOrDocumentSnapshot.length === 1 &&
+      is.instance(fieldValuesOrDocumentSnapshot[0], DocumentSnapshot)
+    ) {
+      const documentSnapshot = fieldValuesOrDocumentSnapshot[0];
+
+      fieldValues = [];
+      let hasDocumentId = false;
+
+      for (let fieldOrder of this._fieldOrders) {
+        if (fieldOrder.field === FieldPath._DOCUMENT_ID) {
+          hasDocumentId = true;
+          fieldValues.push(documentSnapshot.ref);
+        } else {
+          let fieldValue = documentSnapshot.get(fieldOrder.field);
+          if (is.undefined(fieldValue)) {
+            throw new Error(
+              `Field '${
+                fieldOrder.field
+              }' is missing in the provided DocumentSnapshot. Please provide a ` +
+                'document that contains values for all specified orderBy() ' +
+                'constraints.'
+            );
+          } else {
+            fieldValues.push(fieldValue);
+          }
+        }
+      }
+
+      if (!hasDocumentId) {
+        // Add implicit sorting by name, using the last specified direction.
+        let lastDirection =
+          this._fieldOrders.length === 0
+            ? directionOperators.ASC
+            : this._fieldOrders[this._fieldOrders.length - 1].direction;
+
+        this._fieldOrders.push(
+          new FieldOrder(FieldPath.documentId(), lastDirection)
+        );
```

Comment:
So I was surprised to find a state mutation buried inside a method called buildPosition_() (from the name I assumed no side-effects). And I'm not sure if this is appropriate, since we're modifying the Query object you're calling startAt() _on_ rather than the one that gets _returned_. I'm not 100% sure if that breaks anything (what happens if you then do an explicit orderBy on the document id?), but in general I don't think `query.startAt(...)` should result in mutating `query`. 

FWIW- this method is also getting pretty huge. I wonder if it'd make sense to create a startAtEndAtHelper_() that might look something like:

```
let fieldValues;
let fieldOrders = this._fieldOrders;
if (args[0] is DocumentSnapshot) {
  // Code to ensure fieldOrders contains document ID...
  fieldValues = this.getFieldValuesFromDocumentSnapshot(args[0], fieldOrders);
} else {
  fieldValues = args;
}

const position = this.buildPosition_(fieldValues);

return new Query(..., fieldOrders, position);
```
(sorry this is terrible pseudocode and feel free to choose better method names, but hopefully you get the idea)

Reply:
Hm, so I thought about this PR some more and found more issues. For once, there is the issue you pointed out (modifying the immutable state), and there there is this rabbit-hole that Jonny once warned me about: `query.where(...).startAfter(doc)` will break in unexpected ways (at least for the casual user).

So, in order to be able to append `__name__` to where() queries, I ported the logic from the Mobile Clients. I used this to split up the method into three smaller ones:

- Extracting the field values
- Computing the implicit order by
- Generation of the Protobuf message based on the fields and the order by constraint

Unlike the Mobile SDKs, I am rejecting calls to `where()` once a user has specified a cursor to not invalidate it. This matches the behavior of the `orderBy()` API.